src/main.cairo

```diff
@@ -1,32 +1,179 @@
+}
+
+@constructor
+func constructor{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    naming_address: felt, 
+    pricing_address: felt
+) {
+    naming_contract.write(naming_address);
+    pricing_contract.write(pricing_address);
+    return ();
 }
 
+// @notice Get the status of renewals for a domain
+// @param domain Domain to get status of
+// @param user User to get status of
+@view
+func is_renewing{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt, 
+    user: felt
+) -> (res: felt) {
+    let (res) = renews.read(user, domain);
+    return (res,);
+}
+
+// @notice Get the status of an upgrade vote
+// @param upgrade Upgrade to get status of
+// @param user User to get status of
+@view
+func has_voted_upgrade{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    user: felt,
+    upgrade: felt, 
+) -> (res: felt) {
+    let (res) = voted_upgrade.read(user, upgrade);
+    return (res,);
+}
+
+//
+// Externals
+//
+
+// @notice Activate or deactivate renewals for a domain
+// @param domain Domain to activate or deactivate renewals for
+// @dev callable by owner of domain
 @external
-func increase_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
-    amount: felt
+func toggle_renewals{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt,
 ) {
-    with_attr error_message("Amount must be positive. Got: {amount}.") {
-        assert_nn(amount);
+    alloc_locals;
+    let (caller) = get_caller_address();
+    let (contract) = naming_contract.read();
+    let (owner) = Naming.domain_to_address(contract, 1, cast(new(domain), felt*));
+    with_attr error_message("Caller is not owner of domain") {
+        assert owner = caller;
     }
 
-    let (res) = balance.read();
-    balance.write(res + amount);
+    let (prev_renew) = renews.read(caller, domain);
+    renews.write(caller, domain, 1 - prev_renew);
+
+    ToggledRenewal.emit(domain, caller, 1 - prev_renew);
+
+    return ();
+} 
+
+// @notice Renew a domain
+// @param domain Domain to renew
+// @dev callable by anyone, but will only renew if renewals are activated for domain
+@external
+func renew{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    root_domain: felt
+) {
+    alloc_locals;
+    let (naming) = naming_contract.read(); 
+    let (owner) = Naming.domain_to_address(naming, 1, cast(new(root_domain), felt*));
+    let (can_renew) = renews.read(owner, root_domain);
+
+    with_attr error_message("Owner has not activated renewals for this domain") {
+        assert_not_zero(can_renew);
+    }
+
+    let (expiry) = Naming.domain_to_expiry(naming, 1, cast(new(root_domain), felt*));
+    let (block_timestamp) = get_block_timestamp();
+    let time_remaining = expiry - block_timestamp;
+    with_attr error_message("Domain is not set to expire within a month") {
+        assert_le(time_remaining, 86400 * 30);
```

Comment:
I think it would be nice to be able to renew a domain even if it has expired (even though this should not happen if our bot is running correctly). To do so we should check expiry <= block_timestamp + 86400 * 30, because otherwise expiry-block_timestamp would be negative and thus overflow, the assert would fail.